src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs

```diff
@@ -2491,12 +2508,31 @@ public override BoundNode VisitConditionalAccess(BoundConditionalAccess node)
                     SetUnreachable();
                 }
 
-                VisitRvalue(node.AccessExpression);
-                Join(ref this.State, ref savedState);
+                // We want to preserve StateWhenNotNull from accesses in the same "chain":
+                // a?.b(out x)?.c(out y); // expected to preserve StateWhenNotNull from both ?.b(out x) and ?.c(out y)
+                // but not accesses in nested expressions:
+                // a?.b(out x, c?.d(out y)); // expected to preserve StateWhenNotNull from a?.b(out x, ...) but not from c?.d(out y)
+                BoundExpression cursor = node.AccessExpression;
+                while (cursor is BoundConditionalAccess innerCondAccess)
```

Comment:
>cursor [](start = 23, length = 6)

Not really sure what cursor is supposed to mean.

Reply:
This lets us traverse conditional accesses within a chain and keep the null-conditional state within them in State while discarding null-conditional state in nested conditional accesses.

Reply:
No, I get what it's doing. Just not the name.

---
In reply to: [592814927](https://github.com/dotnet/roslyn/pull/51567#discussion_r592814927) [](ancestors = 592814927)

Reply:
Ah, I used it to mean "a unit of the thing I am traversing". We can rename it to anything else, maybe just `expr`, if you prefer.